File: codewars/sudoku/sudoku.py
errors = 0


def sudoku_solver(grid):
    global errors
    linesX = None

    def solve():
        global errors
        loops = 0
        numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
        linesX = grid
        linesY = []
        squares = []

        for i in range(9):
            yLine = []
            for j in range(9):
                yLine.append(linesX[j][i])
            linesY.append(yLine)

        for l in range(0, 9, 3):
            for k in range(0, 9, 3):
                square = []
                for i in range(l, l + 3):
                    for j in range(k, k + 3):
                        if linesX[i][j] != 0:
                            square.append(linesX[i][j])
                squares.append(square)

        block = None
        solved = False
        while not solved:
            loops += 1
            for line in linesX:
                if 0 in line:
                    solved = False
                    break
                else:
                    solved = True

            for i in range(9):
                if i % 3 == 0:
                    block = i
                for j in range(9):
                    print("yo")
                    neededNums = [num for num in numbers if num not in linesX[i]]
                    if linesX[i][j] == 0:
                        neededNum = [num for num in neededNums if num not in linesY[j]]
                        possibleNums = [
                            num
                            for num in neededNum
                            if num not in squares[j // 3 + block]
                        ]
                        print(possibleNums)
                        if len(possibleNums) == 1:
                            linesX[i][j] = possibleNums[0]
                            linesY[j][i] = possibleNums[0]
                            squares[j // 3 + block].append(possibleNums[0])
                        elif len(possibleNums) == 2 and errors == 0 and loops > 70:
                            linesX[i][j] = possibleNums[0]
                            linesY[j][i] = possibleNums[0]
                            squares[j // 3 + block].append(possibleNums[0])
                        elif len(possibleNums) == 2 and loops > 70:
                            linesX[i][j] = possibleNums[1]
                            linesY[j][i] = possibleNums[1]
                            squares[j // 3 + block].append(possibleNums[0])
        for i in range(9):
            for j in range(1, 10):
                if linesX[i].count(j) > 1:
                    errors += 1

    solve()
    while errors > 0:
        solve()
    return linesX


print(
    sudoku_solver(
        [
            [0, 0, 6, 1, 0, 0, 0, 0, 8],
            [0, 8, 0, 0, 9, 0, 0, 3, 0],
            [2, 0, 0, 0, 0, 5, 4, 0, 0],
            [4, 0, 0, 0, 0, 1, 8, 0, 0],
            [0, 3, 0, 0, 7, 0, 0, 4, 0],
            [0, 0, 7, 9, 0, 0, 0, 0, 3],
            [0, 0, 8, 4, 0, 0, 0, 0, 6],
            [0, 2, 0, 0, 5, 0, 0, 8, 0],
            [1, 0, 0, 0, 0, 2, 5, 0, 0],
        ]
    )
)

File: codewars/sudoku/sudoku.js
const sudoku = (grid) => {
	const numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
	let linesX = grid
	const linesY = []
	const squares = []

	for (let i = 0; i < 9; i++) {
		const yLine = []
		for (let j = 0; j < 9; j++) {
			yLine.push(grid[j][i])
		}
		linesY.push(yLine)
	}

	for (let l = 0; l < 9; l += 3) {
		for (let k = 0; k < 9; k += 3) {
			const square = []
			for (let i = l; i < l + 3; i++) {
				for (let j = k; j < k + 3; j++) {
					if (grid[i][j] !== 0) {
						square.push(grid[i][j])
					}
				}
			}
			squares.push(square)
		}
	}
	let block
	let solved = false
	while (solved === false) {
		for (let line of linesX) {
			if (line.includes(0)) {
				solved = false
				break
			} else {
				solved = true
			}
		}
		for (let i = 0; i < 9; i++) {
			if (i % 3 === 0) {
				block = i
			}
			for (let j = 0; j < 9; j++) {
				let neededNums = numbers.filter((num) => !linesX[i].includes(num))
				if (linesX[i][j] === 0) {
					let neededNum = neededNums.filter((num) => !linesY[j].includes(num))
					const possibleNums = neededNum.filter((num) => !squares[Math.floor(j / 3 + block)].includes(num))
					if (possibleNums.length === 1) {
						linesX[i][j] = possibleNums[0]
						linesY[j][i] = possibleNums[0]
						squares[Math.floor(j / 3 + block)].push(possibleNums[0])
					}
				}
			}
		}
	}
	return linesX
}

const puzzle = [
	[5, 3, 0, 0, 7, 0, 0, 0, 0],
	[6, 0, 0, 1, 9, 5, 0, 0, 0],
	[0, 9, 8, 0, 0, 0, 0, 6, 0],
	[8, 0, 0, 0, 6, 0, 0, 0, 3],
	[4, 0, 0, 8, 0, 3, 0, 0, 1],
	[7, 0, 0, 0, 2, 0, 0, 0, 6],
	[0, 6, 0, 0, 0, 0, 2, 8, 0],
	[0, 0, 0, 4, 1, 9, 0, 0, 5],
	[0, 0, 0, 0, 8, 0, 0, 7, 9],
]
console.log(sudoku(puzzle))

const letterRev = (str) => {
	let outputArr = []
	for (let word of str.split(" ")) {
		outputArr.push(word.split("").reverse().join(""))
	}
	return outputArr.join(" ")
}

const scream = (num) => {
	if (num < 1) {
		return "crickets"
	} else {
		return "lol" + "ol".repeat(num - 1)
	}
}

console.log(scream(10))

const createMap = (size) => {
	let output = ""
	for (let i = 0; i < size; i++) {
		for (let j = 0; j < size; j++) {
			if (i % 2 === 0) {
				output += "#0"
			} else {
				output += "0#"
			}
		}
		output += "\n"
	}
	return output
}
console.log(createMap(3))
